lib/msf/core/db_manager/session.rb

```diff
@@ -170,54 +115,129 @@ def report_session(opts)
+      end
+      mod_detail = ::Mdm::Module::Detail.find_by_fullname(mod_fullname)
+      mod_name = mod_detail.name
+
+      vuln_info = {
+        exploited_at: Time.now.utc,
+        host: host,
+        info: "Exploited by #{mod_fullname} to create Session #{s.id}",
+        name: mod_name,
+        refs: mod_detail.refs.map { |ref| ref.name },
+        workspace: wspace,
+      }
 
-    mod = framework.modules.create(session.via_exploit)
+      port    = session.exploit_datastore["RPORT"]
+      service = (port ? host.services.find_by_port(port.to_i) : nil)
 
-    if session.via_exploit == "exploit/multi/handler" and session.exploit_datastore['ParentModule']
-      mod_fullname = session.exploit_datastore['ParentModule']
-      mod_name = ::Mdm::Module::Detail.find_by_fullname(mod_fullname).name
-    else
-      mod_name = mod.name
-      mod_fullname = mod.fullname
-    end
+      vuln_info[:service] = service if service
+
+      vuln = framework.db.report_vuln(vuln_info)
+
+      attempt_info = {
+        host: host,
+        module: mod_fullname,
+        refs: mod_detail.refs,
+        service: service,
+        session_id: s.id,
+        timestamp: Time.now.utc,
+        username: session.username,
+        vuln: vuln,
+        workspace: wspace,
+      }
 
-    vuln_info = {
-      :host => host.address,
-      :name => mod_name,
-      :refs => mod.references,
-      :workspace => wspace,
-      :exploited_at => Time.now.utc,
-      :info => "Exploited by #{mod_fullname} to create Session #{s.id}"
+      framework.db.report_exploit_success(attempt_info)
+
+      vuln
     }
+  end
 
-    port    = session.exploit_datastore["RPORT"]
-    service = (port ? host.services.find_by_port(port.to_i) : nil)
+  def create_mdm_session_from_session(opts)
+    ::ActiveRecord::Base.connection_pool.with_connection {
+      raise ArgumentError.new("Invalid :session, expected Msf::Session") unless opts[:session].kind_of? Msf::Session
+      session = opts[:session]
+      wspace = opts[:workspace] || find_workspace(session.workspace)
+      h_opts = { }
+      h_opts[:host]      = normalize_host(session)
+      h_opts[:arch]      = session.arch if session.respond_to?(:arch) and session.arch
+      h_opts[:workspace] = wspace
+      host = find_or_create_host(h_opts)
+      sess_data = {
+        datastore: session.exploit_datastore.to_h,
+        desc: session.info,
+        host_id: host.id,
+        last_seen: Time.now.utc,
+        local_id: session.sid,
+        opened_at: Time.now.utc,
+        platform: session.platform,
+        port: session.session_port,
+        routes: [],
+        stype: session.type,
+        via_exploit: session.via_exploit,
+        via_payload: session.via_payload,
+      }
 
-    vuln_info[:service] = service if service
+      if session.exploit_task and session.exploit_task.record
+        session_task = session.exploit_task.record
+        if session_task.class == Mdm::Task
```

Comment:
Why is this using `class ==` instead of `session_task.is_a? Mdm::Task`.  Is something wrong with subclasses of `Mdm::Task`?